syntax = "proto3";
package ortools_vrp;

message Matrix {
  repeated float time = 2 [packed=true];
  repeated float distance = 3 [packed=true];
  repeated float value = 4 [packed=true];
}

message TimeWindow {
  int64 start = 1;
  int64 end = 2;
}

message TimeTable {
  repeated TimeWindow timewindows = 1;
  uint32 duration = 2;
  uint32 setup_duration = 3;
  float late_multiplier = 4;
  string id = 5;
}

message Service {
  repeated uint32 quantities = 1;
  uint32 priority = 2;
  repeated int32 vehicle_indices = 3;
  uint32 matrix_index = 4;
  uint32 timetable_index = 5;
  string type = 6;
  string id = 7;
  repeated uint32 setup_quantities = 8;
  uint32 additional_value = 9;
  string alternative_id = 10;
}

message Rest {
  repeated TimeWindow time_windows = 1;
  uint64 duration = 2;
}

message Capacity {
  int32 limit = 1;
  float overload_multiplier = 2;
  bool counting = 3;
}

message Vehicle {
  repeated Capacity capacities = 3;
  TimeWindow time_window = 4;
  repeated Rest rests = 5;
  float cost_fixed = 6;
  float cost_distance_multiplier = 7;
  float cost_time_multiplier = 8;
  float cost_waiting_time_multiplier = 9;
  uint32 matrix_index = 10;
  int32 start_index = 11;
  int32 end_index = 12;
  int64 duration = 13;
  bool force_start = 14;
  float cost_late_multiplier = 15;
  int32 day_index = 16;
  uint32 value_matrix_index = 17;
  float cost_value_multiplier = 18;
}

message Relation {
  string type = 1;
  repeated string linked_ids = 2;
  int32 lapse = 3;
}

message Problem {
  repeated Vehicle vehicles = 3;
  repeated Service services = 4;
  repeated Matrix matrices = 5;
  repeated Relation relations = 6;
  repeated TimeTable timetables = 7;
}
